modules/gdscript/gdscript.cpp

```diff
@@ -1394,38 +1394,40 @@ String GDScript::debug_get_script_name(const Ref<Script> &p_script) {
 thread_local GDScript::UpdatableFuncPtr GDScript::func_ptrs_to_update_thread_local;
 GDScript::UpdatableFuncPtr *GDScript::func_ptrs_to_update_main_thread = &func_ptrs_to_update_thread_local;
 
-List<GDScript::UpdatableFuncPtrElement>::Element *GDScript::_add_func_ptr_to_update(GDScriptFunction **p_func_ptr_ptr) {
+void GDScript::_add_func_ptr_to_update(GDScriptFunction **p_func_ptr_ptr, List<UpdatableFuncPtrElement>::Element **r_element_ptr) {
 	MutexLock lock(func_ptrs_to_update_mutex);
 
 	List<UpdatableFuncPtrElement>::Element *result = func_ptrs_to_update_elems.push_back(UpdatableFuncPtrElement());
 
 	{
 		MutexLock lock2(func_ptrs_to_update_thread_local.mutex);
 		result->get().element = func_ptrs_to_update_thread_local.ptrs.push_back(p_func_ptr_ptr);
-		result->get().mutex = &func_ptrs_to_update_thread_local.mutex;
+		result->get().element_ptr = r_element_ptr;
+		result->get().thread_mutex = &func_ptrs_to_update_thread_local.mutex;
+		result->get().script_mutex = &func_ptrs_to_update_mutex;
+		*r_element_ptr = result;
 
 		if (likely(func_ptrs_to_update_thread_local.initialized)) {
-			return result;
+			return;
 		}
 
 		func_ptrs_to_update_thread_local.initialized = true;
 	}
 
 	func_ptrs_to_update.push_back(&func_ptrs_to_update_thread_local);
-
-	return result;
 }
 
 void GDScript::_remove_func_ptr_to_update(List<UpdatableFuncPtrElement>::Element *p_func_ptr_element) {
 	// None of these checks should ever fail, unless there's a bug.
 	// They can be removed once we are sure they never catch anything.
 	// Left here now due to extra safety needs late in the release cycle.
 	ERR_FAIL_NULL(p_func_ptr_element);
-	MutexLock lock(func_ptrs_to_update_thread_local.mutex);
-	ERR_FAIL_NULL(p_func_ptr_element->get().element);
-	ERR_FAIL_NULL(p_func_ptr_element->get().mutex);
-	MutexLock lock2(*p_func_ptr_element->get().mutex);
+	ERR_FAIL_NULL(p_func_ptr_element->get().thread_mutex);
+	ERR_FAIL_NULL(p_func_ptr_element->get().script_mutex);
+	MutexLock lock(*p_func_ptr_element->get().thread_mutex);
+	MutexLock lock2(*p_func_ptr_element->get().script_mutex);
```

Comment:
TSAN complains with this:
```
WARNING: ThreadSanitizer: lock-order-inversion (potential deadlock) (pid=3350594)
  Cycle in lock order graph: M0 (0x7b7000023bc8) => M1 (0x7f85314aa0e0) => M0

  Mutex M1 acquired here while holding mutex M0 in main thread:
    ...
    #7 GDScript::_add_func_ptr_to_update(GDScriptFunction**, List<GDScript::UpdatableFuncPtrElement, DefaultAllocator>::Element**) modules/gdscript/gdscript.cpp:1403 (godot.linuxbsd.editor.dev.x86_64.san+0x6203a31)
    ...

  Mutex M0 acquired here while holding mutex M1 in main thread:
    ...
    #7 GDScript::_remove_func_ptr_to_update(List<GDScript::UpdatableFuncPtrElement, DefaultAllocator>::Element*) modules/gdscript/gdscript.cpp:1428 (godot.linuxbsd.editor.dev.x86_64.san+0x6203d34)
    ...
```

By commenting/removing 1426 and 1428, it seems to satisfy TSAN, because it seems that the mutex in those lines is already locked.

```suggestion
	ERR_FAIL_NULL(p_func_ptr_element->get().thread_mutex);
	MutexLock lock(*p_func_ptr_element->get().thread_mutex);
```

Reply:
I confirm that your fix still works after this.